test/Data.purs

```diff
@@ -402,13 +387,52 @@ instance ToData FType' where
 -- https://github.com/purescript/purescript/issues/2975
 -- https://github.com/purescript/documentation/blob/master/errors/CycleInDeclaration.md
 instance Arbitrary CType where
-  arbitrary = fix $ \arb -> arb
+  arbitrary =
+    (frequency <<< wrap) $
+      (0.25 /\ pure C0)
+        :| (0.25 /\ (C1 <$> arbitrary))
+          List.: (0.25 /\ (C2 <$> arbitrary <*> arbitrary))
+          List.: (0.25 /\ (C3 <$> arbitrary <*> arbitrary <*> arbitrary))
+          List.: List.Nil
 
 instance Arbitrary EType where
   arbitrary = genericArbitrary
 
 instance Arbitrary DType where
-  arbitrary = fix $ \arb -> arb
+  arbitrary = toDType <$> arbitrary <*> chooseInt 0 15
```

Comment:
Is there a reason you don't use a similar approach as you did in CType? Seems quite nice and declarative there...

Reply:
At first I had something like 

```haskell
genArbitraryDType :: Int -> Gen DType
genArbitraryDType n = 
  if n<=0 then 
    nonRecursive
  else 
      do
      choose <- chooseInt 0 2
      if choose < 2 then 
        nonRecursive
      else 
        D1 <$> genArbitraryDType (n-1)
  where 
    nonRecursive :: Gen DType
    nonRecursive = (frequency <<< wrap) $
      (0.5 /\ (D0 <$> arbitrary <*> arbitrary <*> arbitrary))
      :| (0.5 /\ (D2 <$> arbitrary))
      List.: List.Nil
```

Then I got a suggestion in slack to use the correspondence of types to write it, I think that it looks better and it's easier to understand, but also think that the old way is more declarative. I don't have really a preference here.

Reply:
The real problem here is the recursivity of the `DType` it prevents us to write it like : 

```pureScript
arbitrary = (frequency <<< wrap) $
      (0.45 /\ (D0 <$> arbitrary <*> arbitrary <*> arbitrary))
      :| (0.45 /\ (D2 <$> arbitrary))
      :| (0.1 /\ (D1 <$> arbitrary))
      List.: List.Nil
```
And point to https://github.com/purescript/documentation/blob/master/errors/CycleInDeclaration.md

I didn't find the right way to use `fix` here, so that's why I wrote the function `genAribtraryDtype` by hand.

Reply:
the pattern is: `a = f a` can be replaced with `a = fix \_ -> f a`, does that help?

Could something like this work?

```
arbitrary = fix  \arb -> (frequency <<< wrap) $
      (0.45 /\ (D0 <$> arb <*> arb <*> arb))
      :| (0.45 /\ (D2 <$> arb))
      List.: (0.1 /\ (D1 <$> arb))
      List.: List.Nil
```